prestashopv1/plugin/modules/indodana/tools/IndodanaTools.php

```diff
@@ -0,0 +1,166 @@
+        $products = $cart->getProducts();
+        foreach ($products as $product) {
+            $items[] = [
+                'id' => $product['reference'],
+                'name' => $product['name'],
+                'price' => round(self::getPriceWithoutReductionWithoutTax($product)),
+                'quantity' => $product['quantity'],
+                'category'  => IndodanaCommon\IndodanaConstant::DEFAULT_ITEM_CATEGORY,
+            ];
+        }
+
+        return $items;
+    }
+
+    public static function getCustomerDetails($orderId)
+    {
+        $orderDetails = new Order((int) $orderId);
+        $customer = new Customer((int) $orderDetails->id_customer);
+        $invoiceDetails = new Address((int) $orderDetails->id_address_invoice);
+
+        return [
+            'firstName' => $customer->firstname,
+            'lastName' => $customer->lastname,
+            'email' => $customer->email,
+            'phone' => $invoiceDetails->phone
+        ];
+    }
+
+    public static function getBillingAddress($orderId)
+    {
+        $orderDetails = new Order((int) $orderId);
+        $invoiceDetails = new Address((int) $orderDetails->id_address_invoice);
+
+        return [
+            'firstName' => $invoiceDetails->firstname,
+            'lastName' => $invoiceDetails->lastname,
+            'address' => $invoiceDetails->address1 . ' ' . $invoiceDetails->address2,
+            'city' => $invoiceDetails->city,
+            'postalCode' => $invoiceDetails->postcode,
+            'phone' => $invoiceDetails->phone,
+            'countryCode' => Configuration::get('INDODANA_STORE_COUNTRY_CODE')
+        ];
+    }
+
+    public static function getDeliveryAddress($orderId)
+    {
+        $orderDetails = new Order((int) $orderId);
+        $deliveryDetails = new Address((int)($orderDetails->id_address_delivery));
+
+        return [
+            'firstName' => $deliveryDetails->firstname,
+            'lastName' => $deliveryDetails->lastname,
+            'address' => $deliveryDetails->address1 . ' ' . $deliveryDetails->address2,
+            'city' => $deliveryDetails->city,
+            'postalCode' => $deliveryDetails->postcode,
+            'phone' => $deliveryDetails->phone,
+            'countryCode' => Configuration::get('INDODANA_STORE_COUNTRY_CODE')
+        ];
+    }
+
+    public static function getTotalProduct($product)
+    {
+        return round(self::getPriceWithoutReductionWithoutTax($product) * $product['quantity']);
+    }
+
+    /**
+     * sometimes when rounding up/down numbers there will be deficit
+     * so that the total price calculation does not match
+     */
+    public static function getDeficit($product)
+    {
+        $total = self::getTotalProduct($product);
+
+        return $total % 2 ? floor($product['quantity'] / 2) : 0;
+    }
+
+    public static function getPriceWithoutReductionWithoutTax($product)
+    {
+        return $product['price_without_reduction'] / (($product['rate'] + 100) / 100);
```

Comment:
I'm not too sure about this calculation. We'll discuss again later

Reply:
It would be preferable to use some built in API prestashop if any though instead of rough calculation

Reply:
Prestashop 1.6 doesn't provide API to "get_price_without_reduction_without_tax", so we need to calculate it manually